pkg/container/types/array.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"io"
+	"strconv"
+	"strings"
+	"unsafe"
+)
+
+// NOTE: vecf32 and vecf64 in SQL is internally represented using T_array_float32 and T_array_float64.
+// array is used to avoid potential conflicts with the already existing vector class from vectorized execution engine.
+
+const (
+	MaxArrayDimension = 65536
+)
+
+func BytesToArray[T RealNumbers](input []byte) (res []T) {
+	return DecodeSlice[T](input)
+}
+
+func ArrayToBytes[T RealNumbers](input []T) []byte {
+	return EncodeSlice[T](input)
+}
+
+func ArrayToString[T RealNumbers](input []T) string {
+	var buffer bytes.Buffer
+	_, _ = io.WriteString(&buffer, "[")
+	for i, value := range input {
+		if i > 0 {
+			_, _ = io.WriteString(&buffer, ", ")
+		}
+		_, _ = io.WriteString(&buffer, fmt.Sprintf("%v", value))
+	}
+	_, _ = io.WriteString(&buffer, "]")
+	return buffer.String()
+}
+
+func ArraysToString[T RealNumbers](input [][]T) string {
+	strValues := make([]string, len(input))
+	for i, row := range input {
+		strValues[i] = ArrayToString[T](row)
+	}
+	return strings.Join(strValues, " ")
+}
+
+func StringToArray[T RealNumbers](str string) ([]T, error) {
+	input := strings.TrimSpace(str)
```

Comment:
isn't this very slow?   first you trim sapce, then check prefix/suffix, then replaceall.   later split string ...

why cannot we use a loop ...

Reply:
unsafe access to StringIndex did not yield much of a performance improvement.

```
BenchmarkStringToArray
BenchmarkStringToArray/StringToArray-Trim_Split
BenchmarkStringToArray/StringToArray-Trim_Split-10         	1000000000	         0.0000722 ns/op
BenchmarkStringToArray/StringToArray-ForLoop-RuneCast
BenchmarkStringToArray/StringToArray-ForLoop-RuneCast-10   	    4512	    250405 ns/op
BenchmarkStringToArray/StringToArray-ForLoop-Unsafe
BenchmarkStringToArray/StringToArray-ForLoop-Unsafe-10     	    6218	    187644 ns/op
```

<details>
<summary>StringToArrayV3</summary>

```go


// StringToArrayV3 this implementation uses only one forloop without casting string to rune array (via unsafe)
func StringToArrayV3[T RealNumbers](str string) ([]T, error) {
	strLen := len(str)
	strIdx := 0

	// Trim Space from beginning
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// If beginning is not [ it is malformed.
	if unsafeStringAt(str, strIdx) != '[' {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// skip [
	strIdx++

	// we take the subarray from x based on final array dimension.
	x := make([]T, MaxArrayDimension)
	dim := 0
	var err error

	// find the next "," from the strIdx. More like IndexOf(",", startIdx) from java.
	commaPos := indexFrom(str, ",", strIdx)

	for strIdx < strLen && unsafeStringAt(str, strIdx) != ']' {
		if dim == MaxArrayDimension {
			return nil, moerr.NewInternalErrorNoCtx("typeLen is over the MaxVectorLen: %v", MaxArrayDimension)
		}
		// skip space in the numStr, Eg:- "[  1, 2]"
		for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
			strIdx++
		}

		// if there is no "," then break and check for the "]" condition
		if commaPos == -1 {
			break
		}

		// this means that numStr is ""
		if strIdx == commaPos {
			return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
		}

		// convert numStr to T
		x[dim], err = stringToT[T](str[strIdx:commaPos])
		if err != nil {
			return nil, err
		}
		dim++

		// move strIdx to pos after commaPos
		strIdx = commaPos + 1

		// find the next comma pos from strIdx.
		// If there is none, the break condition kicks in and rest of the code checks for
		// "]" condition.
		commaPos = indexFrom(str, ",", strIdx)
	}

	// skip space for case:- [   ]
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// find next occurrence of "]" from strIdx.
	endBracketPos := indexFrom(str, "]", strIdx)
	if endBracketPos == -1 {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// if numStr is ""
	if strIdx == endBracketPos {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// convert numStr to T and incr dim
	x[dim], err = stringToT[T](str[strIdx:endBracketPos])
	if err != nil {
		return nil, err
	}
	dim++

	// set strIdx after endBracket
	strIdx = endBracketPos + 1

	// Trim of the trailing spaces
	for strIdx < strLen && unicode.IsSpace(unsafeStringAt(str, strIdx)) {
		strIdx++
	}

	// If after TrimEnd, there are still data in the str, then it is malformed.
	// Eg:- [1,2,3] 1
	if strIdx != strLen {
		return nil, moerr.NewInternalErrorNoCtx("malformed vector input: %s", str)
	}

	// return subarray of x based on the final dimension.
	return x[:dim], nil
}

// unsafeStringAt used when we want str[idx] without casting str to []rune
func unsafeStringAt(str string, idx int) rune {
	return rune(*(*byte)(unsafe.Pointer((*reflect.StringHeader)(unsafe.Pointer(&str)).Data + uintptr(idx))))
}

// indexFrom Find the next occurrence of substr after start
func indexFrom(str, substr string, start int) int {
	pos := strings.Index(str[start:], substr)
	if pos != -1 {
		pos += start
	}
	return pos
}

// stringToT convert str to T
func stringToT[T RealNumbers](str string) (t T, err error) {
	switch any(t).(type) {
	case float32:
		num, err := strconv.ParseFloat(str, 32)
		if err != nil {
			return t, moerr.NewInternalErrorNoCtx("error while casting %s to %s", str, T_float32.String())
		}
		// FIX: https://stackoverflow.com/a/36391858/1609570
		numf32 := float32(num)
		return *(*T)(unsafe.Pointer(&numf32)), nil
	case float64:
		num, err := strconv.ParseFloat(str, 64)
		if err != nil {
			return t, moerr.NewInternalErrorNoCtx("error while casting %s to %s", str, T_float64.String())
		}
		return *(*T)(unsafe.Pointer(&num)), nil
	default:
		panic(moerr.NewInternalErrorNoCtx("not implemented"))
	}
}


```
</details>

Reply:
@fengttt 
I think we should need [Binary Insert](https://github.com/matrixorigin/matrixone/issues/9955) for better performance.

Based on a quick benchmark, we could get up to 8x performance
```
    # mo       macos   run2: Result: vector dim=1024 vectors inserted=40960 insert/second=2650.0632385796366 Binary
    # mo       macos   run3: Result: vector dim=1024 vectors inserted=40960 insert/second=2701.284438140351 BinaryString
```